repo.go

```diff
@@ -37,12 +37,138 @@ type Interface interface {
 // Charm store references will use the provided parameters.
 // Local references will use the provided path.
 func InferRepository(ref *charm.URL, charmStoreParams NewCharmStoreParams, localRepoPath string) (Interface, error) {
-	switch ref.Schema {
+	dis := NewDispatcher(charmStoreParams, localRepoPath)
+	inferred, err := dis.Infer(ref.Schema)
+	if err != nil {
+		return nil, err
+	}
+	repo, err := inferred.NewRepo()
+	if err != nil {
+		return nil, err
+	}
+	return repo, nil
+}
+
+// Dispatcher supports selecting and performing repo-related behavior
+// based on charm schemas. Behavior selection is independent of the
+// source of the schema (e.g. charm.URL).
+//
+// Recognized schemas are:
+//   cs
+//   local
+type Dispatcher struct {
+	// Factory is the charm repo factory that the dispatcher uses.
+	Factory Factory
+
+	// Handlers maps charm schemas to functions that will be called
+	// in response to the corresponding schema. A missing entry
```

Comment:
I don't understand what a handler does here. What does it mean to "handle" a charm repository?